File: #0838.push-dominoes.cpp
class Solution {
public:
    string pushDominoes(string dominoes) {

        int size = dominoes.size();
        vector<int> L(size, INT_MAX);
        vector<int> R(size, INT_MAX);
        for (int i = 0; i < size; ++i) {
            if (dominoes[i] == 'L') {
                L[i] = 0;
                for (int j = i - 1; j >= 0 && dominoes[j] == '.'; --j) {
                    L[j] = L[j + 1] + 1;
                }
            } else if (dominoes[i] == 'R') {
                R[i] = 0;
                for (int j = i + 1; j < size && dominoes[j] == '.'; ++j) {
                    R[j] = R[j - 1] + 1;
                }
            }
        }

        for (int i = 0; i < size; ++i) {
            if (L[i] < R[i]) {
                dominoes[i] = 'L';
            } else if (L[i] > R[i]) {
                dominoes[i] = 'R';
            }
        }
        return dominoes;
    }
};

File: #0838.push-dominoes.js
/**
 * @param {string} dominoes
 * @return {string}
 */
var pushDominoes = function(dominoes) {
    const N = dominoes.length;
    const dp = new Array(N).fill().map(_ => {
        return { left: Infinity, right: Infinity } 
    });
    for (let last = '', force = 0, i = 0; i < N; i++) {
        if (dominoes[i] != '.') {
            force = 0;
            last = dominoes[i];
        }
        
        if (last == 'R') {
            dp[i].right = force++;
        }
    }
    
    for (let last = '', force = 0, i = N-1; i >= 0; i--) {
        if (dominoes[i] != '.') {
            force = 0;
            last = dominoes[i];
        }
        
        if (last == 'L') {
            dp[i].left = force++;
        }
    }
    
    return dp.map(({ left, right }) => left === right ? '.' : left < right ? 'L' : 'R').join('');
};
